### ISSUE TITLE
How to evaluate validation loss during training.

### ISSUE THREAD

**Issue by Kunaldargan** [IDEA-Research/detrex#373]
Dear Authors, 
Thank you for maintaining an excellent repository and keeping it intuitive for learners.

I wanted to confirm there is a better way to have validation loss during train time.

[How do I compute validation loss during training?](https://github.com/facebookresearch/detectron2/issues/810#top)
#810
Taken inspiration from: https://github.com/facebookresearch/detectron2/issues/810#issuecomment-596194293

```
import copy
from detectron2.engine import HookBase
class ValidationLoss(HookBase):
	def __init__(self, cfg):
		super().__init__()
		self.cfg = copy.deepcopy(cfg)
		## In order to get instances from test data
		#print("self.cfg.dataloader.test.keys():", self.cfg.dataloader.test.keys()) 
		##self.cfg.dataloader.test.keys(): dict_keys(['dataset', 'mapper', 'num_workers', '_target_'])

		self.cfg.dataloader.test.mapper.is_train =  True
		self.data_loader = instantiate(self.cfg.dataloader.test)

	def after_step(self):
		if self.trainer.iter % (self.cfg.train.eval_period // 4) == 0:  # Evaluate every 25% of eval_period
	    		self._compute_validation_loss()

	def _compute_validation_loss(self):
		assert self.trainer.model.training, "[Trainer] model was changed to eval mode!"
		assert torch.cuda.is_available(), "[Trainer] CUDA is required for AMP training!"
		from torch.cuda.amp import autocast

		"""
		If you want to do something with the data, you can wrap the dataloader.
		"""
		total_loss = 0
		num_batches = 0
		loss_sums = {}  # Dictionary to accumulate loss for each component

		with torch.no_grad():
			for data in self.data_loader: #tqdm.tqdm()
				##print(data) ## Contains instances when self.cfg.dataloader.test.mapper.is_train : True
				"""
				If you want to do something with the losses, you can wrap the model.
				"""
				with autocast(enabled=self.trainer.amp):
					loss_dict = self.trainer.model(data)
					if isinstance(loss_dict, torch.Tensor):
						losses = loss_dict
						loss_dict = {"total_loss": loss_dict}
					else:
						losses = sum(loss_dict.values())

					#loss_dict = self.trainer.model(data)
					#print("loss_dict:", loss_dict)

					# Accumulate each loss separately
					for key, loss in loss_dict.items():
						loss_value = loss.item()
						loss_sums[key] = loss_sums.get(key, 0) + loss_value  # Accumulate loss per key
        
					total_loss += losses.item()
					num_batches += 1

			# Compute dataset-wide average losses
			avg_losses = {key: value / num_batches for key, value in loss_sums.items()} if num_batches > 0 else {}

			# Store scalar values
			for key, avg_loss in avg_losses.items():
			    self.trainer.storage.put_scalar(f"validation_{key}", avg_loss)

			avg_val_loss = total_loss / num_batches if num_batches > 0 else 0
			self.trainer.storage.put_scalar("validation_loss", avg_val_loss)

			print(f"Validation Loss Breakdown: {avg_losses}")  # Optional logging

```

Please add validation loss evaluation as a feature. 

Thank you
